### PR TITLE
WIP: Use load/dump exclusively for decode/encode

### PR DESCRIPTION
**Pull request by ericmj**

### CODE REVIEW

File: lib/mongo_ecto.ex [hunk 1]
```diff
@@ -407,23 +407,68 @@ defmodule Mongo.Ecto do
...
-    do: Ecto.Type.load(ObjectID, data, &load/2)
-  def load(Ecto.Date, {date, _time}),
-    do: Ecto.Type.load(Ecto.Date, date, &dump/2)
-  def load(:map, []),
-    do: {:ok, %{}}
+  def load(_type, nil),
+    do: {:ok, nil}
+  # wat
+  def load(:binary_id, <<_::binary-12>> = binary),
+    do: {:ok, ObjectID.encode(binary)}
```
Comment (michalmuskala) [inline lib/mongo_ecto.ex]: This is very strange... Any ideas where it's coming from?

Comment (ericmj) [inline lib/mongo_ecto.ex]: No, hoping @josevalim can tell us.

File: lib/mongo_ecto.ex [hunk 2]
```diff
@@ -407,23 +407,68 @@ defmodule Mongo.Ecto do
...
+  def load(:binary_id, %BSON.ObjectId{value: value}),
+    do: ObjectID.load(value)
+  def load(:binary, %BSON.Binary{binary: value}),
+    do: {:ok, value}
+  def load(Ecto.UUID, %BSON.Binary{binary: value}),
+    do: {:ok, value}
```
Comment (michalmuskala) [inline lib/mongo_ecto.ex]: Is this correct? Shouldn't we encode it in some way?

Comment (ericmj) [inline lib/mongo_ecto.ex]: I assumed both uuid and binary_id was hex encoded at some point. binary_id is definitely covered by tests and works, we can add a test for uuid to verify that it also works.

Comment (michalmuskala) [inline lib/mongo_ecto.ex]: Just checked and it indeed works correctly.

File: lib/mongo_ecto.ex (hunk 1 above)
Comment (josevalim) [inline lib/mongo_ecto.ex]: Certainly not from Ecto. From here I think: https://github.com/michalmuskala/mongodb_ecto/blob/02fe793d04e9c7833259466ba5dae8b58d755b7d/lib/mongo_ecto.ex#L475-L484

File: lib/mongo_ecto.ex (hunk 2 above)
Comment (josevalim) [inline lib/mongo_ecto.ex]: I think this is wrong. The value in the model will be wrong.

File: lib/mongo_ecto.ex (hunk 1 above)
Comment (josevalim) [inline lib/mongo_ecto.ex]: If you return the whole ObjectId in the linked code above, you should be able to remove this clause.

File: lib/mongo_ecto.ex (hunk 2 above)
Comment (ericmj) [inline lib/mongo_ecto.ex]: What is the correct way to handle it?

Comment (josevalim) [inline lib/mongo_ecto.ex]: The value needs to be the UUID string and not a binary. That's all. If the value is a binary, you can call Ecto.UUID.load and you will be good to go.

Comment (ericmj) [inline lib/mongo_ecto.ex]: Apparently mongodb sends the UUID as a string - not as a binary so I believe this is working correctly and tests confirm that.

**Comment by ericmj** [comment]
This is ready for review. /cc @josevalim @michalmuskala

**Comment by josevalim** [comment]
:shipit:

**Comment by ericmj** [comment]
@michalmuskala Can you do a new release? This is a pretty bug imo so would be nice to have in stable.

**Comment by michalmuskala** [comment]
I agree. I'm holding off for now, because there are some issues with 1.1, but I haven't got time to investigate them yet. I'll try to do that as soon as possible, and release.

**Comment by ericmj** [comment]
Can you release for 1.0 and then work on 1.1?

On Wed, Jan 13, 2016 at 7:07 PM, Michał Muskała <notifications@github.com>
wrote:

> I agree. I'm holding off for now, because there are some issues with 1.1,
> but I haven't got time to investigate them yet. I'll try to do that as soon
> as possible, and release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/michalmuskala/mongodb_ecto/pull/58#issuecomment-171382494>
> .
>



-- 
Eric Meadows-Jönsson

**Comment by michalmuskala** [comment]
I released 0.1.3 with restriction for ecto 1.0. I'll try to release 0.2.0 for ecto 1.1 later today.

**Comment by ericmj** [comment]
Awesome, thanks a lot Michal.

**Pull request merged**